Implement analyzer changes for supporting array(subquery) expressions

smakalias opened the pull request:
## Summary of the changes / Why this improves CrateDB

Implement analyzer changes for supporting array(subquery) expressions

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

andreidan commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalysisContext.java:
```diff
@@ -38,6 +38,7 @@
     private final Map<SubqueryExpression, Object> arrayExpressionsChildren = new IdentityHashMap<>();
 
     private boolean hasAggregates;
+    private boolean forceReturnMultipleRows;
```
I'd give this a less "apocalyptic" name - maybe something along the lines of "isReturnMultiplRows".

andreidan commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java:
```diff
@@ -527,6 +528,17 @@ protected Symbol visitInPredicate(InPredicate node, ExpressionAnalysisContext co
+        protected Symbol visitArraySubQueryExpression(ArraySubQueryExpression node, ExpressionAnalysisContext context) {
+            context.forceReturnMultipleRows(true);
+            Symbol argument = process(node.subqueryExpression(), context);
+            context.forceReturnMultipleRows(false);
```
I think it's save to leave it true for the entire operation no?

smakalias replied:
what if there is another array(subquery) operation in the where clause or in another nested query? Would it still be safe?

andreidan replied:
This is the context for this expression. The nested expressions would have their own analysis. @mfussenegger do you have anything to add?

andreidan commented:
@smakalias this looks fairly good, left a couple of comments.

mfussenegger commented on sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalysisContext.java:
```diff
@@ -47,6 +48,14 @@ public boolean hasAggregates() {
     }
 
+    void forceReturnMultipleRows(boolean forceReturnMultipleRows) {
+        this.forceReturnMultipleRows = forceReturnMultipleRows;
```
I don't understand why we have to introduce a second facility for this. Can't we use the existing `isArrayComparisonChild` functionality, but adapt it to fit both cases?

smakalias commented:
@mfussenegger can you please check again? I did use the same functionality as you suggested.

smakalias commented:
@andreidan please review 
- implemented changed according to team input and recommendations
- added unit test for analyser and also integration test

Note: I will update documentation on the feature branch - after merging this one

andreidan commented:
@smakalias can you please squash commit 2-3-4 as they're achieving the final output (doing/undoing/redoing the same bit of work). Other than that this is looking good.

mfussenegger commented on sql/src/test/java/io/crate/integrationtests/ArraySubQueryIntegrationTest.java:
```diff
@@ -0,0 +1,57 @@
+    public void testSubQueryInProjectionList() throws Exception {
+        setup.setUpCharacters();
+
+        execute("select id, name, array(select id from characters where female = true) as array_females from characters order by id");
```
We could probably change the queries to use sys tables to avoid the (relatively expensive) create table - inserts.

smakalias replied:
done
